presto-main/src/main/java/com/facebook/presto/sql/planner/CanonicalPlanGenerator.java

```diff
@@ -88,6 +100,141 @@ public Optional<PlanNode> visitPlan(PlanNode node, Context context)
         return Optional.empty();
     }
 
+    @Override
+    public Optional<PlanNode> visitJoin(JoinNode node, Context context)
+    {
+        if (strategy == DEFAULT) {
+            return Optional.empty();
+        }
+        if (node.getType().equals(JoinNode.Type.RIGHT)) {
+            return visitJoin(node.flipChildren(), context);
+        }
+        List<PlanNode> sources = new ArrayList<>();
+        ImmutableList.Builder<RowExpression> allFilters = ImmutableList.builder();
+        ImmutableList.Builder<JoinNode.EquiJoinClause> criterias = ImmutableList.builder();
+        Stack<JoinNode> stack = new Stack<>();
+
+        stack.push(node);
+        while (!stack.empty()) {
+            JoinNode top = stack.pop();
+            top.getCriteria().forEach(criteria -> criterias.add(criteria));
```

Comment:
nit: `top.getCriteria().forEach(criterias::add);`

Reply:
I so much want to add this to maven-checks...